src/librustdoc/html/render.rs

```diff
@@ -1343,12 +1343,17 @@ impl Context {
             } else {
                 let mut url = repeat("../").take(cx.current.len())
                                            .collect::<String>();
-                if let Some(&(ref names, _)) = cache().paths.get(&it.def_id) {
+                if let Some(&(ref names, ty)) = cache().paths.get(&it.def_id) {
                     for name in &names[..names.len() - 1] {
                         url.push_str(name);
                         url.push_str("/");
                     }
-                    url.push_str(&item_path(it));
+                    let target_name = names.last().unwrap();
+                    if let ItemType::Module = ty {
+                        write!(url, "{}/index.html", target_name).unwrap();
+                    } else {
+                        write!(url, "{}.{}.html", ty.to_static_str(), target_name).unwrap();
```

Comment:
This seems duplicated with `item_path`, perhaps the two could be refactored to share code?